refactor: restructure the implementation

smusali opened the pull request:
refactor: restructure the implementation

This commit is also introducing huge data handling, which is needed mainly for CloudTrail.

Ref: #2
Semver: major

---

test: add unit test coverage

Initializing TAP and adding unit test cases to cover all the cases

Ref: #3
Semver: patch

---

feat(config): use @logdna/env-config for configuration

Replace current configuration with [@logdna/env-logdna](https://www.npmjs.com/package/@logdna/env-config)

Ref: #2
Semver: major

---

refactor(logger): introduce @logdna/logger as a logger

Replacing the current flushing mechanism with @logdna/logger

Ref: #2
Semver: minor

---

docs: update README and Contribution Guide and add Code of Conduct

Include the same `CONTRIBUTING.md` and `CODE_OF_CONDUCT.md` we have been using for other repos.

Ref: #8
Semver: patch

Commit:
refactor: restructure the implementation

This commit is also introducing huge data handling which is needed mainly for CloudTrail

Ref: #2
Semver: major

matt-march commented on .eslintrc:
```diff
@@ -1,207 +1,225 @@
 {
```
Should we just use the [logdna eslint package](https://github.com/logdna/eslint-config-logdna) instead of this `.eslintrc`?

smusali replied:
Done!

matt-march commented on index.js:
```diff
@@ -1,255 +1,270 @@
+  return {...eventData, line: formatMessage(eventData)}
+}
+
+// Extra the Main, Meta, and Time Data
```
`Extra` -> `Extract`?

smusali replied:
Done!

matt-march commented on index.js:
```diff
@@ -1,255 +1,270 @@
+  }, []) : [{line: JSON.stringify(data), meta, timestamp}]
+}
 
 // Gett the Logs from File
```
`Gett` -> `Get`

smusali replied:
Done!

matt-march commented on index.js:
```diff
@@ -1,255 +1,270 @@
+  const {Records, logFiles, ...meta} = data
+  const timestamp = meta && meta.eventTime ? (new Date(meta.eventTime)).getTime() : undefined
+  data = Records || logFiles || data
+  return Array.isArray(data) ? data.reduce((lines, entry) => {
```
This feels more like a `map` operation (like it was before) than a `reduce` one.

smusali replied:
Done!

matt-march reviewed (commented):
It would be great to split up the `index.js` into multiple files to make it more understandable/maintainable. This would also make it easier to add tests.

Nit: would also be easier to review the main changes if the `eslint` changes were in their own commit.

evanlucas commented on index.js:
```diff
@@ -1,255 +1,269 @@
+'use strict'
```
This really needs tests. Especially since there is a material change here

smusali replied:
Working on it

smusali commented:
Thanks for the feedback, guys @evanlucas @matt-march! I have decided to make changes one-by-one as listed [here](https://github.com/logdna/logdna-s3/issues)

Commit:
test: add unit test coverage

Initializing TAP and adding unit test cases to cover all the cases

Ref: #3
Semver: patch

mdeltito commented on .gitignore:
```diff
@@ -13,4 +13,9 @@ node_modules
+# No TAP Artifacts
+.tap
+coverage
+.nyc_output
```
newline at the end ↩️

mdeltito replied:
on line 2:

package-lock.json should not be excluded through `.gitignore`, either:

1. disable the lockfile with a `.npmrc` or
2. include the lockfile

This project provides an installable artifact, it's not a dependency. So it should really have a lockfile (remove from `gitignore` and commit it)

smusali replied:
Done!

mdeltito commented on .circleci/config.yml:
```diff
@@ -12,14 +12,14 @@ test_build_filters: &test_build_filters
 jobs:
```
not necessarily here in the PR, but we need to get this off circle CI and manual releases

smusali replied:
Gonna be part of #4

mdeltito commented on index.js:
```diff
@@ -1,255 +1,48 @@
+exports.handler = (event, context, callback) => {
+  const conf = config.getConfig(event)
+  , eventData = eventHandler.processEvent(eventHandler.parseEvent(event))
+  , payload = conf.eventlog ? [eventHandler.prepareEvent(eventData)] : []
```
I'm surprised this isn't enforced by our lint config, but one `const` per declaration:

```js
  const conf = config.getConfig(event)
  const eventData = eventHandler.processEvent(eventHandler.parseEvent(event))
  const payload = conf.eventlog ? [eventHandler.prepareEvent(eventData)] : []
```

smusali replied:
Done!

mdeltito commented on index.js:
```diff
@@ -1,255 +1,48 @@
-            batches.push(batch);
-            batch = [];
-            batch_size = 0;
+  if (conf.filelog) {
+    return transformer.getLogs(s3params, (error, lines) => {
```
can you talk a little bit about the scenarios that are being accounted for here with `filelog`, `eventlog`, or _both_? pretty sure I understand but wanted to be sure.

The branch logic is pretty complicated here - we're either extracting lines from a file on s3, or we're getting them from the event itself. Can this just be something like:

```js
let lines
if (isFile) {
  lines = linesFromFile(...)
} else {
  lines = linesFromEvent(...)
}

// treat them the same, batch + flush
```

mdeltito replied:
to put it another way:

as a user, when and why would I set `LOGDNA_EVENTLOG` and/or `LOGDNA_FILELOG`? what specifically are we accounting for with these "modes"?

smusali replied:
We are always extracting lines from S3 buckets in this lambda but event logging means generating logs for the events happening to the targeted S3 buckets; that's what `LOGDNA_EVENTLOG` is for - by default, `LOGDNA_FILELOG=true` and `LOGDNA_EVENTLOG=false`

smusali replied:
@mdeltito, after thinking more about this, I found out that the customers can get the S3 event logs thru CloudWatch and in this lambda, there is no need for such complication; so, I have removed it, which will cause a major version bump but that's fine as long as it keeps the codebase as readable and simple as possible

mdeltito commented on index.js:
```diff
@@ -1,255 +1,48 @@
+
+      const logArrays = payload.concat(transformer.prepareLogs(lines, eventData))
+      const batches = transformer.batchify(logArrays)
+      return async.everySeries(batches, (batch, next) => {
+        setTimeout(() => {
+          return logger.flush(batch, config, next)
+        }, BATCH_INTERVAL)
+      }, callback)
```
why do we need the `setTimeout` and `BATCH_INTERVAL`?

smusali replied:
@mdeltito, I remember I needed to add it since our ingester wasn't stable for sequential ingestions without any interval but I will replace the whole `logger` with our `logger-node`

mdeltito replied:
yeah I think that will help a lot, it already handles buffered flushing (default is 250ms I believe), retries, all that other fun stuff. the only difference being that it's event-based instead of callback based, but that shouldn't change much about this block.

mdeltito commented on lib/config.js:
```diff
@@ -0,0 +1,42 @@
+'use strict'
```
the new env vars in here all need documentation

I'd really recommend using https://github.com/logdna/env-config-node, it's purpose-built for this.

```js
const config = require('./config.js')
config.validateEnvVars()

if ( /*event has hostname*/ ) {
  config.set('hostname', event_host)
}

// etc
```

smusali replied:
I will add this after finishing testing

mdeltito commented on lib/event-handler.js:
```diff
@@ -0,0 +1,54 @@
+
+const formatter = require('./formatter.js')
+
+const parseEvent = (event) => {
```
by default we should use function declarations (which are hoisted)

same is true in other files

```js
module.exports = {
  parseEvent
, prepareEvent
, processEvent
}

function parseEvent() {...}
function prepareEvent() {...}
function processEvent() {...}

smusali replied:
Done for all!

mdeltito commented on lib/event-handler.js:
```diff
@@ -0,0 +1,54 @@
+const formatter = require('./formatter.js')
+
+const parseEvent = (event) => {
+  const key = event.Records[0].s3.object.key
+  event.Records[0].s3.object.key = decodeURIComponent(key.replace(/\+/g, ' '))
```
is it really safe to assume that these nested properties exist? you may want to pull in `get-property` for this
https://github.com/answerbook/stdlib-node/blob/master/lib/object/get-property.js

```js
const key = get(event.Records[0], 's3.object.key')
```

also, why do we need to mutate the event like on line 7?

smusali replied:
Well, our `get-property` is private but I can create similar utility function for this purpose; the reason for mutation is in `AWS SDK S3` response, in the file paths, whitespaces are replaced with `+`; so, it's basically for undoing the process.

mdeltito replied:
yeah I was thinking just copying that into `lib/lang/object`

I would say unless there is a strong reason for undo-ing the encode that S3 does on the file paths _on the event object itself_, we should leave it unmodified. decode it where you need it in a separate variable.

smusali replied:
Well, I need to, since I'm using `s3.object` as a whole - let me implement that utility function - shouldn't be a problem

mdeltito commented on lib/logger.js:
```diff
@@ -0,0 +1,73 @@
+'use strict'
```
curious - why can't we use our node logger here? I may have asked this before but I forget...

it really feels like we should be leaning on that to handle all of this rather than rewriting it. if there is something that prevents us from using it, what is it?

smusali replied:
I will after finishing testing

mdeltito commented on lib/transformer.js:
```diff
@@ -0,0 +1,83 @@
+
+const formatter = require('./formatter.js')
+
+const BATCH_LIMIT = parseInt(process.env.LOGDNA_BATCH_LIMIT) || 25
```
this could be in the config, especially if you switch to `env-config-node`

mdeltito commented on lib/transformer.js:
```diff
@@ -0,0 +1,83 @@
+  apiVersion: '2006-03-01'
+})
+
+const batchify = (logs) => {
+  const batches = []
+  let batch = [], batch_size = 0
+  logs.forEach((log) => {
+    batch.push(log)
+    batch_size += 1
+    if (batch_size >= BATCH_LIMIT) {
+      batches.push(batch)
+      batch = []
+      batch_size = 0
+    }
+  })
+
+  if (batch_size > 0) { batches.push(batch) }
+  return batches
+}
```
https://github.com/answerbook/stdlib-node/blob/master/lib/array/chunk.js

```js
  const size = BATCH_LIMIT
  const out = []
  for (let i = 0; i < arr.length; i += size) {
    out.push(arr.slice(i, i + size))
  }
  return out
```

we should avoid iterating with `forEach`, there's no need for an additional function call when a `for` or `for..of` will do the same, unless you really need the closure (which we do not here)

smusali replied:
Done in `utils`

mdeltito commented on lib/transformer.js:
```diff
@@ -0,0 +1,83 @@
+const extractData = (data) => {
+  const {Records, logFiles, ...meta} = data
+  const ts = meta && meta.eventTime ? (new Date(meta.eventTime)).getTime() : undefined
+  data = Records || logFiles || data
```
avoid reassigning the `data` input to this function

```js
let records = data
if (Records) {
  records = Records
} else if (logFiles) {
  records = logFiles
}
```

where does `logFiles` come from, is this a default field from CloudWatch files or something? 

in general I understand the desire to make this s3 handler generic so that it supports all kinds of data that might be flowing through s3, but it comes at the expense of complexity. just something to consider.

smusali replied:
Done!

mdeltito commented on lib/transformer.js:
```diff
@@ -0,0 +1,83 @@
+      return callback(`Error in Getting ${params.Key} from ${params.Bucket}: ${error}`)
+    }
+
+    data = keyFormat.gz ? zlib.gunzipSync(data.Body) : data.Body
```
avoid reassigning function parameters, better to declare a new variable.

smusali replied:
Fixed!

mdeltito commented on package.json:
```diff
@@ -1,10 +1,59 @@
 {
   "name": "logdna-s3",
-  "version": "1.0.0",
+  "version": "1.1.0",
```
again, we should get this automated and on Jenkins (example from [logdna-bunyan](https://github.com/logdna/logdna-bunyan/commit/a141f9fb504b3ab656d58266b400e4ed26fec5e4)). doesn't need to happen here.


might be worth considering something like [`claudia`](https://github.com/claudiajs/claudia) for the build step. [`claudia pack`](https://github.com/claudiajs/claudia/blob/master/docs/pack.md) provides a standard way to generate a zip file like we're already doing, but does a lot more.

smusali replied:
@mdeltito, we have a separate ticket (#4) for this; do you want me to implement that within this PR or in separate PR after finishing this?

mdeltito replied:
no, as long as we're tracking a ticket to do it after the refactor, that seems good to me

mdeltito commented on package.json:
```diff
@@ -13,7 +62,11 @@
     "aws-sdk": "^2.590.0",
-    "eslint": "^6.7.2"
+    "eslint": "^7.4.0",
+    "eslint-config-logdna": "^2.0.0",
+    "eslint-plugin-node": "^11.1.0",
+    "eslint-plugin-sensible": "^2.2.0",
```
should use the latest `eslint-config-logdna@4.0.2`

you should not need the other two plugins if you are using `eslint-config-logdna`, they are already dependencies.

smusali replied:
Done!

mdeltito commented on index.js:
```diff
@@ -1,255 +1,53 @@
-        config.eventlog = process.env.LOGDNA_EVENTLOG.toLowerCase();
-        config.eventlog = config.eventlog.indexOf('yes') > -1 || config.eventlog.indexOf('true') > -1;
-    }
+function handler(event, context, callback) {
```
these days, the `handler` can be an [async function](https://docs.aws.amazon.com/lambda/latest/dg/nodejs-handler.html), which would then allow you to cleanly use async/await everywhere.

smusali replied:
Will alter this in the next commit where I will replace `logger.flush` with `logger-node`

mdeltito commented on lib/config.js:
```diff
@@ -0,0 +1,31 @@
+    }).join(',')
+  }
+
+  if (event && event.Records && Array.isArray(event.Records)
+    && event.Records[0] && event.Records[0].s3 && event.Records[0].s3.bucket
+    && event.Records[0].s3.bucket.name && !config.hostname
+  ) {
+    config.hostname = event.Records[0].s3.bucket.name.replace(DOT_REGEXP, '_')
+  }
```
I know you plan on changing this with `env-config`, but probably still relevant:
```js
const bucketName = getProperty(event, 'Records.0.s3.bucket.name')
if (bucketName && !config.hostname) {
  config.hostname = bucketName.replace(DOT_REGEXP, '_')
}
```

mdeltito commented on lib/event-handler.js:
```diff
@@ -0,0 +1,67 @@
+}
+
+function parseEvent(event) {
+  if (!hasProperty(event, 'Records')) return undefined
+
+  const records = getProperty(event, 'Records')
+  const checkRecords = Array.isArray(records) && records.length > 0
+  if (!checkRecords) return undefined
```
minor, but you really only need `getProperty` and `hasProperty` when you're dealing with nested props. same is true throughout.

also simpler suggestion:

```js
const records = event.Records
if (!Array.isArray(records) || !records.length) return
```

mdeltito commented on lib/event-handler.js:
```diff
@@ -0,0 +1,67 @@
+  }
+}
+
+function processEvent(eventData) {
```
also minor, but both `parseEvent` and `processEvent` are essentially doing the same thing (creating a new, restructured object from an event/eventData). and as far as I can tell, they are always called together. is there any reason to keep them separate?

feels like this could just be a single function/file called `format-event.js` that does this in 1 step

smusali replied:
combined them in `handleEvent`

mdeltito commented on lib/transformer.js:
```diff
@@ -0,0 +1,111 @@
+  } = data
+
+  let ts = Date.now()
+  if (hasProperty(meta, 'eventTime')) {
+    ts = (new Date(getProperty(meta, 'eventTime'))).getTime()
+  }
```
again, I'd say just go for `meta.eventTime` unless you're dealing with nested properties.

get/hasProperty are useful in the same way [optional chaining](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining) is, but that is not available in node 12

mdeltito commented on lib/transformer.js:
```diff
@@ -0,0 +1,111 @@
+
+    let content = data.Body
+    if (keyFormat.gz) {
+      content = zlib.gunzipSync(content.Body)
```
just a note, if we switch this over to async/await we can more easily use the non-`sync` variant

smusali replied:
@mdeltito, had small difficulty in testing while changing this to `await zlib.gunzip`; therefore, kept only this one as it is but others have changed

mdeltito commented on lib/transformer.js:
```diff
@@ -0,0 +1,111 @@
+
+function prepareLogs(logs, eventData) {
+  return logs.filter((log) => {
+    return log !== '' || !log.line
```
is `log` expected to be either a string _or_ an object? it doesn't look like that is the case below. maybe just:

```js
return !log || !log.line
```

smusali replied:
It's expected to be an object

mdeltito reviewed (commented):
few additional comments but solid improvements in the latest

Commit:
feat(config): use LogDNA's env-config package

Replace current configuration with [@logdna/env-logdna](https://www.npmjs.com/package/@logdna/env-config)

Ref: #2
Semver: major

Commit:
refactor(logger): introduce @logdna/logger as a logger

Replacing current flushing mechanism with @logdna/logger

Ref: #2
Semver: minor

mdeltito commented on index.js:
```diff
@@ -1,255 +1,50 @@
+  }
+
+  let lines
+  try {
+    lines = getLogs(s3params)
+  } catch (e) {
+    return callback(e)
+  }
```
an `async` handler should not use a callback, just let it throw. [reference](https://docs.aws.amazon.com/lambda/latest/dg/nodejs-handler.html#nodejs-handler-async)

smusali replied:
Done!

mdeltito commented on index.js:
```diff
@@ -1,255 +1,50 @@
+    config.set('hostname', s3params.Bucket.replace(DOT_REGEXP, '_'))
+  }
+
+  async.everySeries(batches, (batch, next) => {
+    setTimeout(() => {
+      return flush(batch, config, next)
+    }, config.get('batch-interval'))
+  }, callback)
```
I imagine this will roughly look something like this, and then `lib/logger.js` (and the associated unused deps) can go away:

```js
const {once} = require('events')
const {createLogger} = require('@logdna/logger')

async function handler(...) {
  ...

  const options = {...}
  const logger = createLogger(config.get('key'), options)
  logger.on('error', console.error)
  
  for (const batch of batches) {
    for (const msg of batch) {
      logger.log(msg)
    }
  }

  await once(logger, 'cleared')
}

smusali replied:
Done!

mdeltito commented on lib/transformer.js:
```diff
@@ -0,0 +1,143 @@
+  })
+}
+
+async function getLogs(params, callback) {
```
no need for a callback here either, with an `async` function you'll want to throw errors or return the result

smusali replied:
Done!

smusali commented:
@mdeltito, implementation of the stuff is done; now fixing/updating the current and adding newer unit tests to have 100% coverage

smusali commented:
@mdeltito @darinspivey @evanlucas @matt-march 

Okay, guys, refactoring is done, and now it's having 100% unit testing coverage. This is ready to get reviewed and roasted! :)

mdeltito commented on index.js:
```diff
@@ -1,255 +1,32 @@
+}
+
+async function handler(event, context) {
+  const eventData = handleEvent(event)
```
might be better to log a message and return early when `eventData` is undefined, or throw in `handleEvent` when the event is not valid

smusali replied:
Instead of logging and returning; I just did throw since it's anomaly

mdeltito commented on index.js:
```diff
@@ -1,255 +1,32 @@
+  const logger = createLoggerClient(s3params.Bucket)
+  const lines = await getLogs(s3params)
+  logger.on('error', console.error)
+  const logs = prepareLogs(lines, eventData)
```
`prepareLogs` can return undefined if `lines` is invalid, which will then throw on line 25. should `prepareLogs` always return an array instead?

smusali replied:
Great idea!

mdeltito commented on lib/event-handler.js:
```diff
@@ -0,0 +1,50 @@
+  }
+
+  let timestamp = Date.now()
+  if (record.eventTime) {
+    timestamp = (new Date(record.eventTime)).getTime()
+  }
```
not sure if the date format is well-defined here, but if not it would be safer to avoid `NaN` values:

```js

if (record.eventTime) {
  const eventTimestamp = (new Date(record.eventTime)).getTime()
  if (!isNaN(eventTimestamp)) timestamp = eventTimestamp
}

smusali replied:
Well, it will always parse per [docs](https://docs.aws.amazon.com/lambda/latest/dg/with-s3.html) but let me add for the sake of safety

mdeltito commented on lib/logger.js:
```diff
@@ -0,0 +1,56 @@
+  }
+
+  config.set('tags', tags)
+  let hostname = config.get('hostname')
+  if (!hostname && bucket) {
+    hostname = bucket.replace(BUCKET_NAME_REGEX, '_')
+  }
+
+  if (hostname) {
+    config.set('hostname', hostname.replace(HOSTNAME_REGEX, ''))
+  }
```
any reason this needs to be here? feels like it should be done outside of this function, in `index.js`. instantiating the logger shouldn't affect `config`, it should just use it.

mdeltito commented on lib/transformer.js:
```diff
@@ -0,0 +1,155 @@
+    try {
+      content = zlib.gunzipSync(content)
+    } catch (error) {
+      err = new Error(`Error in Unzipping ${params.Bucket}/${params.Key}: ${error}`)
```
we don't need to include the `params` or the original `error` in the message, including it in `error.meta` is sufficient (here and below)

smusali replied:
I know but I also want to have a self-explanatory error message as well but there is no harm in doing that, no, @mdeltito?

mdeltito replied:
when you interpolate other data into the error message (including the original error), you end up with a really long string that can end up just being noise. example:

```
Error in Unzipping long-bucket-name/really-ultra-super-long-object-key-with/slashes/and-maybe-special-characters: Error: the original error message that we actually care about
```

with a short/descriptive error message + the metadata attached, all of the info is presented in a much more readable format. It's not something I'd block on, just a general best-practice.

smusali replied:
Done!

mdeltito commented on test/unit/index.js:
```diff
@@ -0,0 +1,145 @@
+// to work with (they can never be updated). For this, we will trust that env-config
+// will do its job and read from the envrionment properly, but we will set the
+// config's keys manually in these tests.
+config.validateEnvVars()
+config.set('ingestion-key', 'required-for-anything-to-work')
```
`ingestion-key` should be required in `config.js`, and then swap these two lines (or use `process.env.INGESTION_KEY = 'abc213'` at the top)

mdeltito commented on lib/config.js:
```diff
@@ -0,0 +1,56 @@
+, Config
+    .string('https-proxy')
+    .desc('A secure (https://) proxy URL to pass through before going to LogDNA')
+, Config
+    .string('ingestion-key')
+    .desc('LogDNA Ingestion Key to stream the logs from files')
```
this should be required, correct?

mdeltito commented on test/unit/index.js:
```diff
@@ -0,0 +1,145 @@
+config.validateEnvVars()
+config.set('ingestion-key', 'required-for-anything-to-work')
+
+test('handler is the main method in lambda functions', async (t) => {
```
these tests don't appear to be testing what they say, other than that the logger is invoked. we're missing tests that validate the requests to the ingestion endpoint (you can use `nock` for this). It doesn't have to be too exhaustive but would be good to validate that the proper data is included in the logger request.

smusali replied:
Got it! Working on introducing `nock` as well

mdeltito replied:
👍  yeah the `@logdna/logger` should be a good guide to follow, again just need some basic tests that intercept the request and validate that the right data is being sent to our endpoint. This test file can be moved into `test/integration/index.js`

esatterwhite commented on LICENSE:
```diff
@@ -1,6 +1,6 @@
 MIT License
 
-Copyright (c) 2019 LogDNA, Inc.
+Copyright (c) 2021 LogDNA, Inc.
```
This should be the year of initial release more so than the current year

smusali commented on lib/transformer.js:
```diff
@@ -0,0 +1,158 @@
+
+  let data
+  try {
+    data = await s3.getObject(params).promise()
```
@mdeltito @esatterwhite @darinspivey 

can anyone help me in mocking this? Mocking `s3.getObject` is simple but while testing on `AWS`, I have seen it requires `.promise()` too but having difficulty in testing

Thanks a lot in advance!

darinspivey replied:
Sure, I'll take a look.

darinspivey replied:
@smusali I fixed the issue in a new commit, so sync up with origin.  After reading [the SDK docs](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#getObject-property), I couldn't find any reference to `.promise()`, in fact, all of these except 1 are callback-style.  Using build-style methods (chaining) for API calls makes it *really* hard/yucky to test, so we should stay away from those anyway.

The fix that's up in that commit changes the following:
* Use `promisify(s3.getObject)` and export that
* Call the promisified function from `exports` to facilitate testing
* Fixes the test to facilitate mocking - DO NOT destructure exported properties that you wish to mock, or else you can't mock them properly.
* Change the mock function to `async` to match the code
* Change the assertion to a `t.rejects` which is the proper way to test promise rejections (if you use try/catch, your assertion may not run if it doesn't throw due to a bug!)

I had some other comments regarding the `Error` itself, but I'll comment separately for that.

smusali replied:
Thanks, @darinspivey! Let's get this fixed first and I will modify the rest!

darinspivey commented:
Sorry if we talked about this already, but there doesn't seem to be any true integration tests going on in this project, is that right?  I bet testing S3 like this is hard, but it's very strange to me to see all of the tests in `unit` that feel like they go in `integration` since this is entire projects works with a third-party service.

Is there any plans to try and get true integration tests going?

darinspivey commented on .circleci/config.yml:
```diff
@@ -12,14 +12,14 @@ test_build_filters: &test_build_filters
   test:
     docker:
-      - image: circleci/node:10
+      - image: circleci/node:12
```
Again, sorry if we talked about this.  Why not Jenkins here?  We shouldn't use Circle.  Fine to punt this to another PR if that's desired.

mdeltito replied:
I had suggested doing Jenkins in a separate PR, there's a lot going on in here. that way it's one more PR, then let jenkins cut the next release

darinspivey commented on index.js:
```diff
@@ -1,255 +1,51 @@
+  config.validateEnvVars()
+  const eventData = handleEvent(event)
+  if (!eventData) {
+    const error = new Error(`Cannot parse the event: ${JSON.stringify(event)}`)
```
Do we have to stringify here? I think it's ok to have the main `message` be more general, and let the `meta` speak for itself.  Otherwise, what's the point of putting the event in `meta`? :)

smusali replied:
Nope - done!

darinspivey commented on lib/constants.js:
```diff
@@ -0,0 +1,17 @@
+'use strict'
+
+module.exports = {
+  DEFAULT_HTTP_ERRORS: [
```
Naming.  What's "default" about these?  Most of these (other than ENOTFOUND) are connection-based HTTP errors.

Are there any from this list that need to be added?  https://github.com/logdna/logger-node/blob/main/lib/logger.js#L52-L59

smusali replied:
Removed completely

darinspivey commented on lib/transformer.js:
```diff
@@ -0,0 +1,159 @@
+  try {
+    data = await exports.getObject(params)
+  } catch (error) {
+    err = new Error(`Error in Getting the Object: ${error}`)
+    err.meta = {error, params}
+    throw err
```
When making your own error, consider adding more context.  Seeing just this error in the logs wouldn't tell us where the problem is.  By using `meta` we can give more context.  Also, it's not necessary to include the `error` in the `message` string when it's going into `meta`.  If you look at the current test for this, the error messages are a bit redundant, and that's why.  How about something like this:

```suggestion
    err = new Error('Error during getLogs')
    err.meta = {
      message: 's3.getObjects threw an error'
    , error
    , params
    }
    throw err
```

smusali replied:
Done!

darinspivey commented on lib/transformer.js:
```diff
@@ -0,0 +1,159 @@
+    try {
+      content = JSON.parse(content)
+    } catch (error) {
+      err = new Error(`Error in Parsing the JSON Data from the Object: ${error}`)
```
Ditto on the new suggested style for making these errors

smusali replied:
Done!

darinspivey commented on lib/transformer.js:
```diff
@@ -0,0 +1,159 @@
+    try {
+      content = zlib.gunzipSync(content)
+    } catch (error) {
+      err = new Error(`Error in Unzipping the Object: ${error}`)
```
Ditto on the new suggested style for making these errors.  No need to put `error` here if it's in the meta.

smusali replied:
Done!

darinspivey commented on lib/transformer.js:
```diff
@@ -0,0 +1,159 @@
+
+  if (!hasProperty(data, 'Body')) {
+    err = new Error('Corrupted data returned from the object')
+    err.meta = {params}
```
The error's `message` is fine, but shouldn't the meta add additional info about missing `Body`?  Or is this obfuscated for security reasons?

smusali replied:
yea, `data` may contain some unwanted data and for now, I don't have anything extra to add for this - we can enhance once requested

smusali commented:
> Sorry if we talked about this already, but there doesn't seem to be any true integration tests going on in this project, is that right? I bet testing S3 like this is hard, but it's very strange to me to see all of the tests in `unit` that feel like they go in `integration` since this is entire projects works with a third-party service.
> 
> Is there any plans to try and get true integration tests going?

@darinspivey, there are plans for e2e testing which is what I have been recently doing manually but we will automate that too.

darinspivey commented on test/integration/index.js:
```diff
@@ -0,0 +1,286 @@
+config.validateEnvVars()
+nock.disableNetConnect()
+
+test('handler is the main method in lambda functions', async (t) => {
```
The description doesn't really match what this is doing.

smusali replied:
Done!

darinspivey commented on test/integration/index.js:
```diff
@@ -0,0 +1,286 @@
+    })
+
+    const getObject = s3.getObject
+    s3.getObject = function({
```
This needs to be in sync with the other tests.  It's probably an `async` function at a minimum

smusali replied:
Done!

darinspivey commented on test/integration/index.js:
```diff
@@ -0,0 +1,286 @@
+    nock(buildLoggerURL(config))
+      .post('', (body) => {
+        const numProps = Object.keys(body).length
+        t.strictEqual(numProps, 2, 'Number of request body properties')
```
Seems like this test is more for the logger, but it's hard to say what the intention is.  I see there's another comment on this.

darinspivey replied:
I'll leave this topic alone now.  I think just making the description a little better would help.

smusali commented:
@darinspivey @mdeltito 2 things left to say it's good to go:
1. `promisify(s3.getObject)` works for testing purposes but while testing on `AWS Lambda`, it errors out with this message:
```
{
    "errorType": "Error",
    "errorMessage": "Error in Getting the S3 Object",
    "meta": {
        "error": {
            "errorType": "TypeError",
            "errorMessage": "this.makeRequest is not a function",
            "stack": [
                "TypeError: this.makeRequest is not a function",
                "    at Object.svc.<computed> (/var/runtime/node_modules/aws-sdk/lib/service.js:685:23)",
                "    at internal/util.js:297:30",
                "    at new Promise (<anonymous>)",
                "    at Object.getObject (internal/util.js:296:12)",
                "    at getLogs (/var/task/lib/transformer.js:90:26)",
                "    at Runtime.handler (/var/task/index.js:42:23)",
                "    at Runtime.handleOnce (/var/runtime/Runtime.js:66:25)"
            ]
        },
        "params": {
            "Bucket": "johnnytest2",
            "Key": "AWSLogs/o-4dvyqz27un/267460048499/CloudTrail/ap-south-1/2021/04/23/267460048499_CloudTrail_ap-south-1_20210423T1810Z_dH0LrfaLorvZKHfh.json.gz"
        }
    },
    "stack": [
        "Error: Error in Getting the S3 Object",
        "    at getLogs (/var/task/lib/transformer.js:92:11)",
        "    at async Runtime.handler (/var/task/index.js:42:17)"
    ]
}
```
2. There are 2 weird test issues [here](https://app.circleci.com/pipelines/github/logdna/logdna-s3/87/workflows/92fec27b-c117-40f0-a748-6fb8d2c117fe/jobs/92) that don't happen while testing locally.

Will let you guys know once they are fixed!

evanlucas commented:
@smusali you might need to bind it, maybe something like `promisify(s3.getObject.bind(s3))`

smusali commented:
> @smusali you might need to bind it, maybe something like `promisify(s3.getObject.bind(s3))`

This worked! 🎉 thanks, @evanlucas!

smusali commented:
@darinspivey @mdeltito ready to get reviewed again

mdeltito commented on lib/transformer.js:
```diff
@@ -0,0 +1,159 @@
+      }
+
+      if (eventData) {
+        opts.app = formatObjectKey(eventData.file)
```
just wanted to confirm this is correct, and that the `app` should be set to the key/file. what does the current implementation set for `app` and `file` on the lines?

reason being that it seems like `app` will be really long if it includes the full object key

smusali replied:
on the server-side, `_app` gets trimmed, `_file` becomes full file path (which is the object key) and `_app` becomes just a file name - I wanted to use `file` field for this but in that case, `logger` client automatically denotes `app` as `'default'` but this one works as expected

mdeltito reviewed (dismissed):
one question but LGTM, nice work

darinspivey commented on lib/utils.js:
```diff
@@ -0,0 +1,91 @@
+'use strict'
```
I'm not going to block this at the 11th hour (sorry I missed this before), but we should stop making `utils.js` files.  They become a dumping ground of random functions which makes finding functions or browsing the project difficult.  Instead we should try to [organize functions by what they do](https://logdna.blogin.co/posts/99035) (see "No more utils").  This file in particular would benefit from `stdlib`.  We should expose that package to the public so we don't have to duplicate the code here.  When we do that, we can split this file up into something better organized.

smusali replied:
@darinspivey, yup - this is in my mind, we can do reorganization in the next patch release after this

Commit:
docs: update README and Contribution Guide and add Code of Conduct

Include the same `CONTRIBUTING.md` and `CODE_OF_CONDUCT.md` we have been using for other repos.

Ref: #8
Semver: patch

smusali commented on README.md:
```diff
@@ -2,62 +2,65 @@
 
```
@tanberry, can you review the contents of `.md` files in this Pull Request? Thanks in advance!

tanberry commented on CONTRIBUTING.md:
```diff
@@ -1,15 +1,44 @@
+
+The easiest way to get our attention is to comment on an existing, or open a new
+[issue][].
+
```
@smusali do we want to add a link to Laura's new Community forum? Up to you.... https://community.logdna.com is the link.

smusali replied:
@tanberry, how should I mention this at the bottom? "Please, check out our new Community forum"?

tanberry replied:
Yeah, something like that sounds good. Maybe up on line 26, you could tie it into the existing sentence and say something like "Thank you for your contributions; we appreciate you! We hope you will join us on our community forum for developers, at (https://community.logdna.com](https://community.logdna.com ]."

I think it is good to spell out the URL, so that they see it and hopefully remember it. ;-)

But totally up to you.

smusali replied:
Thanks for the great suggestion, @tanberry! Added!

tanberry commented on README.md:
```diff
@@ -2,62 +2,75 @@
+## How to Use
+### Deploy the Code
 1. Create a [new AWS Lambda function](https://console.aws.amazon.com/lambda/home) and select `Author from scratch`
-2. For the basic information:
```
Step 1 needs a period (.) after `Author from scratch`.

smusali replied:
Done!

tanberry commented on README.md:
```diff
@@ -2,62 +2,75 @@
+
+### Configuration
+#### General Configuration
+If the lambda is being used to stream from gzipped files:
```
"... the Lambda function..."

smusali replied:
Done!

tanberry commented on README.md:
```diff
@@ -2,62 +2,75 @@
+
+**Notes**:
+ * The number of retries is recommended to be set to 0 since retrying lambda execution can result in duplicate logs. It can be modified in `Configuration > Asynchronous invoication`
+ * `Timeout` and `Memory` limit may need to be increased if the file size is too big
```
Use "might" instead of "may". ("May" implies permission, while "might" implies the chances of something happening. :-)

Need a period (.) at end of sentence.

smusali replied:
Done!

tanberry commented on README.md:
```diff
@@ -2,62 +2,75 @@
+2. Set `Memory` limit to, at least, `512 MB`
+
+**Notes**:
+ * The number of retries is recommended to be set to 0 since retrying lambda execution can result in duplicate logs. It can be modified in `Configuration > Asynchronous invoication`
```
Suggested rewrite to avoid using passive voice and avoid use of "since". (Since should usually be replaced with "because".)
"The recommended number of retries is "0" because retrying lambda execution can result in duplicate logs."

Typo in "invocation."

smusali replied:
Done! Thanks for catching the typo, @tanberry!

tanberry commented on README.md:
```diff
@@ -2,62 +2,75 @@
+ * Optional prefix and suffix options are related to object paths within the specified bucket
+
+**Notes**:
+ * This S3 Lambda function and S3 Bucket must be in the same availability zone
```
Needs a period (.) at end of sentence. If a bulleted list is of single words, or even longer phrases, then no period is needed. But if the bulleted item is a full sentence, then it needs a period. ;-)

smusali replied:
Done!

tanberry commented on README.md:
```diff
@@ -2,62 +2,75 @@
+
+**Notes**:
+ * This S3 Lambda function and S3 Bucket must be in the same availability zone
+ * You can specify a bucket only in one trigger and/or S3 Lambda function since a bucket accepts only one subscription but one S3 Lambda function can stream from multiple buckets at the same time
```
"...  in only one trigger..." (reverse the order of "only" and "one".)
Replace "since" with "because". ("Since" is a tricky word, because it can imply time OR causality.)
Needs a period (.) at end pf sentence.

smusali replied:
Done!

tanberry commented on README.md:
```diff
@@ -2,62 +2,75 @@
-* `Node.js.10.x` is the minimum runtime requirement for successfully running this S3 Lambda function
+
+#### Environment Variables
+Set `INGESTION_KEY` variable to your LogDNA ingestion key.
```
Consider this rewrite: "Configure the `INGESTION_KEY` variable to use your LogDNA ingestion key."

smusali replied:
@tanberry, I'd keep this `Set` because of the alignment; I mean, I use `Set` for almost all here and the word `Configuration` itself has been used on top for encapsulating all of these

tanberry commented on README.md:
```diff
@@ -2,62 +2,75 @@
 
```
Can we change this title to "LogDNA Amazon S3" please? :-)

smusali replied:
Updated!

tanberry commented on doc/env.md:
```diff
@@ -0,0 +1,184 @@
+
+### `HOSTNAME`
+
+> Optionally, use alternative host name set through the environment
```
Consider saying "... use an alternative host name configured through the environment"

smusali replied:
Done!

tanberry commented on doc/env.md:
```diff
@@ -0,0 +1,184 @@
+### `USER_AGENT`
+
+> user-agent header value to use while sending logs
+
```
To avoid having a lower-case at beginning of this phrase, let's say "The user-agent header value..."

mdeltito replied:
just a note, this file shouldn't be edited directly - this can be changed in `config.js` and then regenerate

smusali replied:
Yup, @mdeltito, that's how I have just updated!

tanberry commented on doc/v2.md:
```diff
@@ -0,0 +1,20 @@
+## Guide for Migration from v1 to v2
+
+`v2` contains lots of refactoral changes to improve the performance and configuration by introducing [`@logdna/logger`](https://www.npmjs.com/package/@logdna/logger), [`@logdna/env-config`](https://www.npmjs.com/package/@logdna/env-config), and [`eslint-config-logdna`](https://www.npmjs.com/package/eslint-config-logdna). There are several changes in naming and handling environment variables as we ...
```
Suggest that we start the sentence with something like "The v2 release of LogDNA Amazon S3 integration..."or something like that.

smusali replied:
Done!

tanberry commented on doc/v2.md:
```diff
@@ -0,0 +1,20 @@
+
+### Renamed Variables
+* `LOGDNA_KEY` has been replaced with [`INGESTION_KEY`](./env.md#ingestion_key)
+* `LOGDNA_HOSTNAME` has been replaced with [`HOSTNAME`](./env.md#hostname)
```
Not absolutely necessary, but would be better, replace "has been replaced" with "was replaced" throughout this section.

smusali replied:
Done!

tanberry commented on doc/v2.md:
```diff
@@ -0,0 +1,20 @@
+* `LOGDNA_BATCH_LIMIT` has been replaced with [`FLUSH_LIMIT`](./env.md#flush_limit)
+
+### Removed Variables
+* `LOGDNA_FREE_SOCKET_TIMEOUT`, `LOGDNA_MAX_REQUEST_RETRIES`, `LOGDNA_REQUEST_RETRY_INTERVAL`, and `LOGDNA_MAX_REQUEST_TIMEOUT` have been removed since the connection gets handled within the logger client
```
Replace "have been removed" with "were removed" and replace "since" with "because."

Also replace "gets handled" with "is handled."

Add period (.) to end of sentence.

smusali replied:
Done!

tanberry commented on doc/v2.md:
```diff
@@ -0,0 +1,20 @@
+### Removed Variables
+* `LOGDNA_FREE_SOCKET_TIMEOUT`, `LOGDNA_MAX_REQUEST_RETRIES`, `LOGDNA_REQUEST_RETRY_INTERVAL`, and `LOGDNA_MAX_REQUEST_TIMEOUT` have been removed since the connection gets handled within the logger client
+* `LOGDNA_MAX_LINE_LENGTH` has been removed since a line length gets handled in the server-side
+* `LOGDNA_EVENTLOG` and `LOGDNA_FILELOG` have been removed since `v2` streams only the logs from the files within the specified bucket. You can use `AWS CloudTrail Event Logging` to stream Object-level operations from `S3` buckets. Check out [this article](https://docs.amazonaws.cn/en_us/AmazonS3/latest/userguide/enable-cloudtrail-logging-for-s3.html)
```
Same as above with "were replaced".
Add period (.) to end of sentence "Check out this article."
Consider changing "Check out this article." to "For more information, refer to this article."

smusali replied:
Done!

tanberry reviewed (changes requested):
@Samir, all done, take what you like and reject the rest. Let me know if you have any questions!

mdeltito commented on doc/v2.md:
```diff
@@ -0,0 +1,20 @@
+## Guide for Migration from v1 to v2
```
I'd suggest renaming this file, e.g. `migrating-to-v2.md` or something

smusali replied:
Done!

alfeng6 commented on doc/migrating-to-v2.md:
```diff
@@ -0,0 +1,20 @@
+
+### Removed Variables
+* `LOGDNA_FREE_SOCKET_TIMEOUT`, `LOGDNA_MAX_REQUEST_RETRIES`, `LOGDNA_REQUEST_RETRY_INTERVAL`, and `LOGDNA_MAX_REQUEST_TIMEOUT` were removed because the connection is handled within the logger client.
+* `LOGDNA_MAX_LINE_LENGTH` was removed because a line length is handled in the server-side.
```
```suggestion
* `LOGDNA_MAX_LINE_LENGTH` was removed because line length is handled in the server-side.
```

smusali replied:
Done!

alfeng6 commented on doc/migrating-to-v2.md:
```diff
@@ -0,0 +1,20 @@
+
+### Removed Variables
+* `LOGDNA_FREE_SOCKET_TIMEOUT`, `LOGDNA_MAX_REQUEST_RETRIES`, `LOGDNA_REQUEST_RETRY_INTERVAL`, and `LOGDNA_MAX_REQUEST_TIMEOUT` were removed because the connection is handled within the logger client.
+* `LOGDNA_MAX_LINE_LENGTH` was removed because a line length is handled in the server-side.
```
```suggestion
* `LOGDNA_MAX_LINE_LENGTH` was removed because a line length is handled on the server-side.
```

smusali replied:
Done!

alfeng6 commented on doc/migrating-to-v2.md:
```diff
@@ -0,0 +1,20 @@
+* `LOGDNA_EVENTLOG` and `LOGDNA_FILELOG` were removed because `v2` streams only the logs from the files within the specified bucket. You can use `AWS CloudTrail Event Logging` to stream Object-level operations from `S3` buckets. For more information, refer to [this article](https://docs.amazonaws.cn/en_us/AmazonS3/latest/userguide/enable-cloudtrail-logging-for-s3.html).
+
+### New Variables
+* [`PROXY`](./env.md#proxy), [`HTTPS_PROXY`](./env.md#https_proxy), and [`HTTP_PROXY`](./env.md#http_proxy) to provide proxy support
```
```suggestion
* [`PROXY`](./env.md#proxy), [`HTTPS_PROXY`](./env.md#https_proxy), and [`HTTP_PROXY`](./env.md#http_proxy) provide proxy support
```

smusali replied:
Done!

alfeng6 commented on doc/migrating-to-v2.md:
```diff
@@ -0,0 +1,20 @@
+
+### New Variables
+* [`PROXY`](./env.md#proxy), [`HTTPS_PROXY`](./env.md#https_proxy), and [`HTTP_PROXY`](./env.md#http_proxy) to provide proxy support
+* [`USER_AGENT`](./env.md#user_agent) to support custom `user-agent` overrides
```
```suggestion
* [`USER_AGENT`](./env.md#user_agent) supports custom `user-agent` overrides
```

smusali replied:
Done!

alfeng6 reviewed (changes requested):
Added a few small comments on grammar

smusali commented:
Resolved all; I think, it's good to go, right, @mdeltito @alfeng6 @tanberry @darinspivey?
